File: internal/drivers/trendmicro.go
package drivers

import (
	"bytes"
	"context"
	"encoding/json"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/nxadm/tail"
	"github.com/rophy/av-scanner/internal/cache"
	"github.com/rophy/av-scanner/internal/config"
)

var (
	// TrendMicro DS Agent log patterns
	// Sample: 2025-11-21 13:53:06.726130: [ds_am/4] | [SCTRL] (0000-0000-0000, /home/ubuntu/xxxx.file) virus found: 2, act_1st=2, act_2nd=255, act_1st_error_code=0 | scanctrl_vmpd_module.cpp:1538:scanctrl_determine_send_dispatch_result | F7E01:1784DB:4451::
	tmVirusFoundRegex = regexp.MustCompile(`\([^,]+,\s*([^)]+)\)\s*virus found:`)
)

type TrendMicroDriver struct {
	config config.DriverConfig
	logger *slog.Logger
	cache  *cache.DetectionCache
	ctx    context.Context
	cancel context.CancelFunc
}

func NewTrendMicroDriver(cfg config.DriverConfig, logger *slog.Logger, detectionCache *cache.DetectionCache) *TrendMicroDriver {
	ctx, cancel := context.WithCancel(context.Background())
	return &TrendMicroDriver{
		config: cfg,
		logger: logger.With("driver", "trendmicro"),
		cache:  detectionCache,
		ctx:    ctx,
		cancel: cancel,
	}
}

// Start begins the background log watcher
func (d *TrendMicroDriver) Start() error {
	if _, err := os.Stat(d.config.RTSLogPath); err != nil {
		d.logger.Warn("RTS log file not accessible, background watcher not started", "path", d.config.RTSLogPath)
		return nil
	}

	go d.watchLog()
	d.logger.Info("Background log watcher started", "path", d.config.RTSLogPath)
	return nil
}

// Stop stops the background log watcher
func (d *TrendMicroDriver) Stop() {
	d.cancel()
}

func (d *TrendMicroDriver) watchLog() {
	t, err := tail.TailFile(d.config.RTSLogPath, tail.Config{
		Follow:    true,
		ReOpen:    true,
		Poll:      true,
		MustExist: true,
		Location:  &tail.SeekInfo{Offset: 0, Whence: os.SEEK_END},
	})
	if err != nil {
		d.logger.Error("Failed to start log tailer", "error", err)
		return
	}
	defer t.Cleanup()

	for {
		select {
		case <-d.ctx.Done():
			t.Stop()
			return
		case line := <-t.Lines:
			if line == nil || line.Err != nil {
				continue
			}
			d.processLogLine(line.Text)
		}
	}
}

func (d *TrendMicroDriver) processLogLine(line string) {
	// Check for virus detection: (trace-id, /path/to/file) virus found: N
	if matches := tmVirusFoundRegex.FindStringSubmatch(line); matches != nil {
		filePath := strings.TrimSpace(matches[1])
		absPath, err := filepath.Abs(filePath)
		if err != nil {
			absPath = filePath
		}

		d.cache.Add(absPath, &cache.Detection{
			FilePath:  filePath,
			Status:    "infected",
			Signature: "virus",
			Raw:       line,
		})
		d.logger.Debug("Cached virus detection", "path", absPath)
	}
}

func (d *TrendMicroDriver) Engine() config.EngineType {
	return config.EngineTrendMicro
}

func (d *TrendMicroDriver) Config() config.DriverConfig {
	return d.config
}

func (d *TrendMicroDriver) RTSWatch(filePath string, opts WatchOptions) (*ScanResult, error) {
	startTime := time.Now()
	fileID := filepath.Base(filePath)

	absPath, err := filepath.Abs(filePath)
	if err != nil {
		absPath = filePath
	}

	timeout := time.After(opts.Timeout)
	ticker := time.NewTicker(50 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-timeout:
			d.logger.Warn("RTS watch timeout", "filePath", filePath)
			return &ScanResult{
				Status:    StatusClean,
				Engine:    d.Engine(),
				Phase:     PhaseRTS,
				FilePath:  filePath,
				FileID:    fileID,
				Timestamp: time.Now(),
				Duration:  time.Since(startTime).Milliseconds(),
				Raw:       map[string]bool{"timeout": true},
			}, nil

		case <-ticker.C:
			if cached, found := d.cache.Get(absPath); found {
				status := StatusClean
				if cached.Status == "infected" {
					status = StatusInfected
				}

				return &ScanResult{
					Status:    status,
					Engine:    d.Engine(),
					Signature: cached.Signature,
					Phase:     PhaseRTS,
					FilePath:  filePath,
					FileID:    fileID,
					Timestamp: time.Now(),
					Duration:  time.Since(startTime).Milliseconds(),
					Raw:       map[string]string{"logEntry": cached.Raw},
				}, nil
			}
		}
	}
}

func (d *TrendMicroDriver) ManualScan(filePath string) (*ScanResult, error) {
	startTime := time.Now()
	fileID := filepath.Base(filePath)

	// dsa_scan --target <file> --json
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(d.config.Timeout)*time.Millisecond)
	defer cancel()

	cmd := exec.CommandContext(ctx, d.config.ScanBinaryPath, "--target", filePath, "--json")
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else if ctx.Err() == context.DeadlineExceeded {
			return nil, ctx.Err()
		} else {
			return nil, err
		}
	}

	output := stdout.String()
	d.logger.Debug("Manual scan completed", "exitCode", exitCode, "output", output)

	status, signature := d.parseManualScanOutput(output, exitCode)

	return &ScanResult{
		Status:    status,
		Engine:    d.Engine(),
		Signature: signature,
		Phase:     PhaseManual,
		FilePath:  filePath,
		FileID:    fileID,
		Timestamp: time.Now(),
		Duration:  time.Since(startTime).Milliseconds(),
		Raw: map[string]interface{}{
			"exitCode": exitCode,
			"stdout":   output,
			"stderr":   stderr.String(),
		},
	}, nil
}

func (d *TrendMicroDriver) parseManualScanOutput(output string, exitCode int) (ScanStatus, string) {
	// Try JSON parsing first (dsa_scan --json output format)
	var jsonResult struct {
		TraceID           string `json:"traceID"`
		NumOfFileScanned  int    `json:"numOfFileScanned"`
		NumOfFileSkipped  int    `json:"numOfFileSkipped"`
		NumOfFileInfected int    `json:"numOfFileInfected"`
		TimeElapse        float64 `json:"timeElapse"`
		ErrorCode         int    `json:"errorCode"`
		InfectedFiles     []struct {
			FileName    string `json:"fileName"`
			MalwareName string `json:"malwareName"`
		} `json:"infectedFiles"`
	}

	if err := json.Unmarshal([]byte(output), &jsonResult); err == nil {
		// Check if file was skipped (not actually scanned)
		if jsonResult.NumOfFileSkipped > 0 && jsonResult.NumOfFileScanned == 0 {
			d.logger.Warn("File was skipped by scanner", "skipped", jsonResult.NumOfFileSkipped)
			return StatusError, ""
		}

		// Check for infected files
		if jsonResult.NumOfFileInfected > 0 || len(jsonResult.InfectedFiles) > 0 {
			signature := ""
			if len(jsonResult.InfectedFiles) > 0 {
				signature = jsonResult.InfectedFiles[0].MalwareName
			}
			return StatusInfected, signature
		}

		// File was scanned and is clean
		if jsonResult.NumOfFileScanned > 0 {
			return StatusClean, ""
		}

		// No files scanned, treat as error
		return StatusError, ""
	}

	// Text-based fallback
	lowerOutput := strings.ToLower(output)
	if strings.Contains(lowerOutput, "infected") ||
		strings.Contains(lowerOutput, "virus") ||
		strings.Contains(lowerOutput, "malware") {
		return StatusInfected, ""
	}

	if exitCode == 0 {
		return StatusClean, ""
	}
	return StatusError, ""
}

func (d *TrendMicroDriver) CheckHealth() (*EngineHealth, error) {
	health := &EngineHealth{
		Engine:    d.Engine(),
		LastCheck: time.Now(),
	}

	if _, err := os.Stat(d.config.RTSLogPath); err != nil {
		health.Healthy = false
		health.Error = "RTS log not accessible: " + d.config.RTSLogPath
		return health, nil
	}

	health.Healthy = true
	return health, nil
}

func (d *TrendMicroDriver) GetInfo() EngineInfo {
	return EngineInfo{
		Engine:              d.Engine(),
		Available:           true,
		RTSEnabled:          true,
		ManualScanAvailable: true,
	}
}

File: src/drivers/trendmicro.ts
import fs from 'fs';
import path from 'path';
import { Tail } from 'tail';
import { AntivirusDriver } from './base';
import {
  EngineHealth,
  EngineInfo,
  LogEntry,
  TrendMicroConfig,
  UnifiedResult,
  WatchOptions,
} from '../types';
import { Logger } from '../utils/logger';

const DEFAULT_TIMEOUT = 60000;
const DEFAULT_POLL_INTERVAL = 100;

/**
 * Trend Micro DS Agent Driver
 *
 * Interface:
 * - RTS Log: watches log file for SCTRL entries
 * - Manual Scan: executes dsa_scan binary
 */
export class TrendMicroDriver extends AntivirusDriver {
  private tmConfig: TrendMicroConfig;

  constructor(config: TrendMicroConfig, logger: Logger) {
    super(config, logger);
    this.tmConfig = config;
  }

  async rtsWatch(
    filePath: string,
    options: WatchOptions = { timeout: DEFAULT_TIMEOUT, pollInterval: DEFAULT_POLL_INTERVAL }
  ): Promise<UnifiedResult> {
    const startTime = Date.now();
    const fileId = path.basename(filePath);

    // DS Agent RTS is always on - we must watch the log for SCTRL entries
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        tail.unwatch();
        this.logger.warn(`RTS watch timeout for ${filePath}`);
        resolve(
          this.createUnifiedResult(
            filePath,
            fileId,
            'clean',
            'rts',
            startTime,
            undefined,
            { timeout: true, reason: 'No SCTRL entry found within timeout' }
          )
        );
      }, options.timeout);

      const tail = new Tail(this.tmConfig.rtsLogPath, {
        fromBeginning: false,
        follow: true,
      });

      tail.on('line', (line: string) => {
        const entry = this.parseLogEntry(line);
        if (entry && this.pathMatches(entry.filePath, filePath)) {
          clearTimeout(timeout);
          tail.unwatch();

          this.logger.info('RTS detected file scan', {
            filePath,
            status: entry.status,
            signature: entry.signature,
          });

          resolve(
            this.createUnifiedResult(
              filePath,
              fileId,
              entry.status,
              'rts',
              startTime,
              entry.signature,
              { logEntry: line }
            )
          );
        }
      });

      tail.on('error', (error: Error) => {
        clearTimeout(timeout);
        tail.unwatch();
        this.logger.error('Tail error on DS Agent log', { error });
        reject(error);
      });
    });
  }

  async manualScan(filePath: string): Promise<UnifiedResult> {
    // RTS-only mode - manual scan not supported
    // The file should be scanned by host RTS when written to scan directory
    const startTime = Date.now();
    const fileId = path.basename(filePath);

    return this.createUnifiedResult(
      filePath,
      fileId,
      'error',
      'manual',
      startTime,
      undefined,
      { error: 'Manual scan not supported in RTS-only mode' }
    );
  }

  protected parseLogEntry(logLine: string): LogEntry | null {
    // DS Agent SCTRL log format:
    // 2025-11-21 13:11:31.654744: [ds_am/4] | [SCTRL] (0000-0000-0000, /path/to/file) virus found: 2 ...
    // 2025-11-21 13:11:31.654744: [ds_am/4] | [SCTRL] (0000-0000-0000, /path/to/file) failed: 3
    // 2025-11-21 13:11:31.654744: [ds_am/4] | [SCTRL] (0000-0000-0000, /path/to/file) clean

    const sctrlMatch = logLine.match(
      /^(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}\.\d+):\s+\[ds_am\/\d+\]\s+\|\s+\[SCTRL\]\s+\([^,]+,\s*([^)]+)\)\s+(.+)$/
    );

    if (!sctrlMatch) {
      return null;
    }

    const [, timestampStr, filePath, result] = sctrlMatch;
    const timestamp = new Date(timestampStr.replace(' ', 'T'));

    // Check for virus found
    const virusMatch = result.match(/virus found:\s*(\d+)/i);
    if (virusMatch) {
      const signatureMatch = result.match(/virus found:\s*\d+\s*(.+)?/i);
      return {
        timestamp,
        filePath: filePath.trim(),
        status: 'infected',
        signature: signatureMatch?.[1]?.trim() || 'Unknown',
        raw: logLine,
      };
    }

    // Check for failure (failed: 3 is not infection, it's scan failure)
    const failedMatch = result.match(/failed:\s*(\d+)/i);
    if (failedMatch) {
      this.logger.warn('SCTRL scan failed', { filePath, errorCode: failedMatch[1] });
      return {
        timestamp,
        filePath: filePath.trim(),
        status: 'error',
        raw: logLine,
      };
    }

    // Check for clean
    if (result.toLowerCase().includes('clean') || result.toLowerCase().includes('ok')) {
      return {
        timestamp,
        filePath: filePath.trim(),
        status: 'clean',
        raw: logLine,
      };
    }

    // Any other SCTRL entry for this file - assume it was processed
    return {
      timestamp,
      filePath: filePath.trim(),
      status: 'clean',
      raw: logLine,
    };
  }

  async checkHealth(): Promise<EngineHealth> {
    try {
      // RTS-only mode - check if log file exists and is readable
      await fs.promises.access(this.tmConfig.rtsLogPath, fs.constants.R_OK);

      return {
        engine: 'trendmicro',
        healthy: true,
        lastCheck: new Date(),
      };
    } catch (error) {
      return {
        engine: 'trendmicro',
        healthy: false,
        lastCheck: new Date(),
        error: `RTS log not accessible: ${this.tmConfig.rtsLogPath}`,
      };
    }
  }

  getInfo(): EngineInfo {
    return {
      engine: 'trendmicro',
      available: true,
      rtsEnabled: true,
      manualScanAvailable: false,
    };
  }

  private pathMatches(logPath: string, targetPath: string): boolean {
    return path.resolve(logPath) === path.resolve(targetPath);
  }
}
